xen/arch/arm/tee/ffa_msg.c

```diff
@@ -122,13 +115,76 @@ int32_t ffa_handle_msg_send2(struct cpu_user_regs *regs)
          src_msg->msg_offset < sizeof(struct ffa_part_msg_rxtx) )
     {
         ret = FFA_RET_INVALID_PARAMETERS;
-        goto out_unlock_tx;
+        goto out;
     }
 
-    ret = ffa_simple_call(FFA_MSG_SEND2, ffa_get_vm_id(src_d) << 16, 0, 0, 0);
+    if ( FFA_ID_IS_SECURE(src_msg->send_recv_id & GENMASK(15,0)) )
+    {
+        /* Message for a secure partition */
+        if ( !ffa_fw_support_fid(FFA_MSG_SEND2) )
+        {
+            ret = FFA_RET_NOT_SUPPORTED;
+            goto out;
+        }
 
-out_unlock_tx:
-    spin_unlock(&src_ctx->tx_lock);
+        ret = ffa_simple_call(FFA_MSG_SEND2, ffa_get_vm_id(src_d) << 16, 0, 0,
+                              0);
+
+        goto out;
+    }
+
+    /* Message for a VM */
+    if ( (src_msg->send_recv_id & GENMASK(15,0)) == 0 )
```

Comment:
The `src_msg` accesses open for a Time-of-Check-Time-of-Use style of attack.

Reply:
Ack, i will fix that using a local variable